File: runtime/c-api/entities/http_client.cpp
#include "http_client.h"
#include "../utils/strings.h"

#ifdef ALT_CLIENT_API

alt::IBaseObject* HttpClient_GetBaseObject(alt::IHttpClient* httpClient) {
    return dynamic_cast<alt::IBaseObject*>(httpClient);
}

void HttpClient_SetExtraHeader(alt::IHttpClient* httpClient, const char* key, const char* value) {
    httpClient->SetExtraHeader(key, value);
}

void MarshalStringDict(alt::IMValueDict* map, const char**& keys, const char**& values, int32_t& size) {
    size = map->GetSize();
    keys = new const char*[size];
    values = new const char*[size];

    auto i = 0;
    for(auto it = map->Begin(); it; it = map->Next())
    {
        auto key = it->GetKey();
        auto value = it->GetValue().As<alt::IMValueString>()->Value();

        auto keyStr = key.c_str();
        auto keySize = key.size();
        auto allocKey = new char[keySize + 1];
        for (auto j = 0; j < keySize; j++) allocKey[j] = keyStr[j];
        allocKey[keySize] = '\0';
        keys[i] = allocKey;

        auto valueStr = value.c_str();
        auto valueSize = value.size();
        auto allocValue = new char[valueSize + 1];
        for (auto j = 0; j < valueSize; j++) allocValue[j] = valueStr[j];
        allocValue[valueSize] = '\0';
        values[i] = allocValue;
        i++;
    }
}

void HttpClient_GetExtraHeaders(alt::IHttpClient* httpClient, const char**& keys, const char**& values, int32_t& size) {
    auto map = httpClient->GetExtraHeaders();
    MarshalStringDict(map.Get(), keys, values, size);
}

void InvokeCallback(const alt::IHttpClient::HttpResponse response, const void* ptr) {
    const char** keys;
    const char** values;
    int32_t size;
    MarshalStringDict(response.headers.Get(), keys, values, size);
    auto delegate = (HttpResponseDelegate_t) ptr;
    delegate(response.statusCode, response.body.c_str(), keys, values, size);
}

void HttpClient_Get(alt::IHttpClient* httpClient, const char* url, /** ClientEvents.HttpResponseModuleDelegate */ HttpResponseDelegate_t callback) {
    httpClient->Get(InvokeCallback, url, (void *) callback);
}

void HttpClient_Head(alt::IHttpClient* httpClient, const char* url, /** ClientEvents.HttpResponseModuleDelegate */ HttpResponseDelegate_t callback) {
    httpClient->Head(InvokeCallback, url, (void *) callback);
}
void HttpClient_Connect(alt::IHttpClient* httpClient, const char* url, const char* body, /** ClientEvents.HttpResponseModuleDelegate */ HttpResponseDelegate_t callback) {
    httpClient->Connect(InvokeCallback, url, body, (void *) callback);
}

void HttpClient_Delete(alt::IHttpClient* httpClient, const char* url, const char* body, /** ClientEvents.HttpResponseModuleDelegate */ HttpResponseDelegate_t callback) {
    httpClient->Delete(InvokeCallback, url, body, (void *) callback);
}

void HttpClient_Options(alt::IHttpClient* httpClient, const char* url, const char* body, /** ClientEvents.HttpResponseModuleDelegate */ HttpResponseDelegate_t callback) {
    httpClient->Options(InvokeCallback, url, body, (void *) callback);
}

void HttpClient_Patch(alt::IHttpClient* httpClient, const char* url, const char* body, /** ClientEvents.HttpResponseModuleDelegate */ HttpResponseDelegate_t callback) {
    httpClient->Patch(InvokeCallback, url, body, (void *) callback);
}

void HttpClient_Post(alt::IHttpClient* httpClient, const char* url, const char* body, /** ClientEvents.HttpResponseModuleDelegate */ HttpResponseDelegate_t callback) {
    httpClient->Post(InvokeCallback, url, body, (void *) callback);
}

void HttpClient_Put(alt::IHttpClient* httpClient, const char* url, const char* body, /** ClientEvents.HttpResponseModuleDelegate */ HttpResponseDelegate_t callback) {
    httpClient->Put(InvokeCallback, url, body, (void *) callback);
}

void HttpClient_Trace(alt::IHttpClient* httpClient, const char* url, const char* body, /** ClientEvents.HttpResponseModuleDelegate */ HttpResponseDelegate_t callback) {
    httpClient->Trace(InvokeCallback, url, body, (void *) callback);
}
#endif

File: api/AltV.Net.Client/Elements/Entities/HttpClient.cs
using System.Runtime.InteropServices;
using AltV.Net.CApi.ClientEvents;
using AltV.Net.Client.Elements.Data;
using AltV.Net.Client.Elements.Interfaces;
using AltV.Net.Elements.Entities;
using AltV.Net.Shared.Utils;

namespace AltV.Net.Client.Elements.Entities
{
    public class HttpClient : BaseObject, IHttpClient
    {
        public IntPtr HttpClientNativePointer { get; }
        public override IntPtr NativePointer => HttpClientNativePointer;

        private static IntPtr GetBaseObjectNativePointer(ICore core, IntPtr httpClientNativePointer)
        {
            unsafe
            {
                return core.Library.Client.HttpClient_GetBaseObject(httpClientNativePointer);
            }
        }

        public HttpClient(ICore core, IntPtr httpClientNativePointer, uint id) : base(core, GetBaseObjectNativePointer(core, httpClientNativePointer), BaseObjectType.HttpClient, id)
        {
            HttpClientNativePointer = httpClientNativePointer;
        }

        [Obsolete("Use Alt.CreateHttpClient instead")]
        public HttpClient(ICore core) : this(core, core.CreateHttpClientPtr(out var id), id)
        {
            core.PoolManager.HttpClient.Add(this);
        }

        public void SetExtraHeader(string key, string value)
        {
            unsafe
            {
                CheckIfEntityExists();
                var keyPtr = MemoryUtils.StringToHGlobalUtf8(key);
                var valuePtr = MemoryUtils.StringToHGlobalUtf8(value);
                Core.Library.Client.HttpClient_SetExtraHeader(HttpClientNativePointer, keyPtr, valuePtr);
                Marshal.FreeHGlobal(keyPtr);
                Marshal.FreeHGlobal(valuePtr);
            }
        }

        public Dictionary<string, string> GetExtraHeaders()
        {
            unsafe
            {
                CheckIfEntityExists();
                int size = 0;
                var keysPtr = IntPtr.Zero;
                var valuesPtr = IntPtr.Zero;
                Core.Library.Client.HttpClient_GetExtraHeaders(HttpClientNativePointer, &keysPtr, &valuesPtr, &size);
                var keys = Core.MarshalStringArrayPtrAndFree(keysPtr, (uint) size);
                var values = Core.MarshalStringArrayPtrAndFree(valuesPtr, (uint) size);
                return keys
                    .Select((e, i) => new KeyValuePair<string, string>(e, values[i]))
                    .ToDictionary(e => e.Key, e => e.Value);
            }
        }

        private HttpResponse GetHttpResponse(int statusCode, string body, IntPtr headerKeys, IntPtr headerValues, int headerSize)
        {
            var keys = Core.MarshalStringArrayPtrAndFree(headerKeys, (uint) headerSize);
            var values = Core.MarshalStringArrayPtrAndFree(headerValues, (uint) headerSize);
            var headers = keys
                .Select((e, i) => new KeyValuePair<string, string>(e, values[i]))
                .ToDictionary(e => e.Key, e => e.Value);
            return new HttpResponse
            {
                Body = body,
                StatusCode = statusCode,
                Headers = headers
            };
        }

        public async Task<HttpResponse> Get(string url)
        {
            CheckIfEntityExists();
            GCHandle handle;
            HttpResponse? data = null;
            var semaphore = new SemaphoreSlim(0, 1);

            unsafe
            {
                void ResolveTask(int statusCode, string body, IntPtr headerKeys, IntPtr headerValues, int headerSize)
                {
                    data = GetHttpResponse(statusCode, body, headerKeys, headerValues, headerSize);
                    semaphore.Release();
                }

                HttpResponseModuleDelegate resolveTask = ResolveTask;
                handle = GCHandle.Alloc(resolveTask);
                var urlPtr = MemoryUtils.StringToHGlobalUtf8(url);
                Core.Library.Client.HttpClient_Get(NativePointer, urlPtr, resolveTask);
                Marshal.FreeHGlobal(urlPtr);
            }

            await semaphore.WaitAsync();
            handle.Free();
            semaphore.Dispose();

            return data.Value;
        }

        public async Task<HttpResponse> Head(string url)
        {
            CheckIfEntityExists();
            GCHandle handle;
            HttpResponse? data = null;
            var semaphore = new SemaphoreSlim(0, 1);

            unsafe
            {
                void ResolveTask(int statusCode, string body, IntPtr headerKeys, IntPtr headerValues, int headerSize)
                {
                    data = GetHttpResponse(statusCode, body, headerKeys, headerValues, headerSize);
                    semaphore.Release();
                }

                HttpResponseModuleDelegate resolveTask = ResolveTask;
                handle = GCHandle.Alloc(resolveTask);
                var urlPtr = MemoryUtils.StringToHGlobalUtf8(url);
                Core.Library.Client.HttpClient_Head(NativePointer, urlPtr, resolveTask);
                Marshal.FreeHGlobal(urlPtr);
            }

            await semaphore.WaitAsync();
            handle.Free();
            semaphore.Dispose();

            return data.Value;
        }

        public async Task<HttpResponse> Connect(string url, string body)
        {
            CheckIfEntityExists();
            GCHandle handle;
            HttpResponse? data = null;
            var semaphore = new SemaphoreSlim(0, 1);

            unsafe
            {
                void ResolveTask(int statusCode, string body, IntPtr headerKeys, IntPtr headerValues, int headerSize)
                {
                    data = GetHttpResponse(statusCode, body, headerKeys, headerValues, headerSize);
                    semaphore.Release();
                }

                HttpResponseModuleDelegate resolveTask = ResolveTask;
                handle = GCHandle.Alloc(resolveTask);
                var urlPtr = MemoryUtils.StringToHGlobalUtf8(url);
                var bodyPtr = MemoryUtils.StringToHGlobalUtf8(body);
                Core.Library.Client.HttpClient_Connect(NativePointer, urlPtr, bodyPtr, resolveTask);
                Marshal.FreeHGlobal(urlPtr);
                Marshal.FreeHGlobal(bodyPtr);
            }

            await semaphore.WaitAsync();
            handle.Free();
            semaphore.Dispose();

            return data.Value;
        }

        public async Task<HttpResponse> Delete(string url, string body)
        {
            CheckIfEntityExists();
            GCHandle handle;
            HttpResponse? data = null;
            var semaphore = new SemaphoreSlim(0, 1);

            unsafe
            {
                void ResolveTask(int statusCode, string body, IntPtr headerKeys, IntPtr headerValues, int headerSize)
                {
                    data = GetHttpResponse(statusCode, body, headerKeys, headerValues, headerSize);
                    semaphore.Release();
                }

                HttpResponseModuleDelegate resolveTask = ResolveTask;
                handle = GCHandle.Alloc(resolveTask);
                var urlPtr = MemoryUtils.StringToHGlobalUtf8(url);
                var bodyPtr = MemoryUtils.StringToHGlobalUtf8(body);
                Core.Library.Client.HttpClient_Delete(NativePointer, urlPtr, bodyPtr, resolveTask);
                Marshal.FreeHGlobal(urlPtr);
                Marshal.FreeHGlobal(bodyPtr);
            }

            await semaphore.WaitAsync();
            handle.Free();
            semaphore.Dispose();

            return data.Value;
        }

        public async Task<HttpResponse> Options(string url, string body)
        {
            CheckIfEntityExists();
            GCHandle handle;
            HttpResponse? data = null;
            var semaphore = new SemaphoreSlim(0, 1);

            unsafe
            {
                void ResolveTask(int statusCode, string body, IntPtr headerKeys, IntPtr headerValues, int headerSize)
                {
                    data = GetHttpResponse(statusCode, body, headerKeys, headerValues, headerSize);
                    semaphore.Release();
                }

                HttpResponseModuleDelegate resolveTask = ResolveTask;
                handle = GCHandle.Alloc(resolveTask);
                var urlPtr = MemoryUtils.StringToHGlobalUtf8(url);
                var bodyPtr = MemoryUtils.StringToHGlobalUtf8(body);
                Core.Library.Client.HttpClient_Options(NativePointer, urlPtr, bodyPtr, resolveTask);
                Marshal.FreeHGlobal(urlPtr);
                Marshal.FreeHGlobal(bodyPtr);
            }

            await semaphore.WaitAsync();
            handle.Free();
            semaphore.Dispose();

            return data.Value;
        }

        public async Task<HttpResponse> Patch(string url, string body)
        {
            CheckIfEntityExists();
            GCHandle handle;
            HttpResponse? data = null;
            var semaphore = new SemaphoreSlim(0, 1);

            unsafe
            {
                void ResolveTask(int statusCode, string body, IntPtr headerKeys, IntPtr headerValues, int headerSize)
                {
                    data = GetHttpResponse(statusCode, body, headerKeys, headerValues, headerSize);
                    semaphore.Release();
                }

                HttpResponseModuleDelegate resolveTask = ResolveTask;
                handle = GCHandle.Alloc(resolveTask);
                var urlPtr = MemoryUtils.StringToHGlobalUtf8(url);
                var bodyPtr = MemoryUtils.StringToHGlobalUtf8(body);
                Core.Library.Client.HttpClient_Patch(NativePointer, urlPtr, bodyPtr, resolveTask);
                Marshal.FreeHGlobal(urlPtr);
                Marshal.FreeHGlobal(bodyPtr);
            }

            await semaphore.WaitAsync();
            handle.Free();
            semaphore.Dispose();

            return data.Value;
        }

        public async Task<HttpResponse> Post(string url, string body)
        {
            CheckIfEntityExists();
            GCHandle handle;
            HttpResponse? data = null;
            var semaphore = new SemaphoreSlim(0, 1);

            unsafe
            {
                void ResolveTask(int statusCode, string body, IntPtr headerKeys, IntPtr headerValues, int headerSize)
                {
                    data = GetHttpResponse(statusCode, body, headerKeys, headerValues, headerSize);
                    semaphore.Release();
                }

                HttpResponseModuleDelegate resolveTask = ResolveTask;
                handle = GCHandle.Alloc(resolveTask);
                var urlPtr = MemoryUtils.StringToHGlobalUtf8(url);
                var bodyPtr = MemoryUtils.StringToHGlobalUtf8(body);
                Core.Library.Client.HttpClient_Post(NativePointer, urlPtr, bodyPtr, resolveTask);
                Marshal.FreeHGlobal(urlPtr);
                Marshal.FreeHGlobal(bodyPtr);
            }

            await semaphore.WaitAsync();
            handle.Free();
            semaphore.Dispose();

            return data.Value;
        }

        public async Task<HttpResponse> Put(string url, string body)
        {
            CheckIfEntityExists();
            GCHandle handle;
            HttpResponse? data = null;
            var semaphore = new SemaphoreSlim(0, 1);

            unsafe
            {
                void ResolveTask(int statusCode, string body, IntPtr headerKeys, IntPtr headerValues, int headerSize)
                {
                    data = GetHttpResponse(statusCode, body, headerKeys, headerValues, headerSize);
                    semaphore.Release();
                }

                HttpResponseModuleDelegate resolveTask = ResolveTask;
                handle = GCHandle.Alloc(resolveTask);
                var urlPtr = MemoryUtils.StringToHGlobalUtf8(url);
                var bodyPtr = MemoryUtils.StringToHGlobalUtf8(body);
                Core.Library.Client.HttpClient_Put(NativePointer, urlPtr, bodyPtr, resolveTask);
                Marshal.FreeHGlobal(urlPtr);
                Marshal.FreeHGlobal(bodyPtr);
            }

            await semaphore.WaitAsync();
            handle.Free();
            semaphore.Dispose();

            return data.Value;
        }

        public async Task<HttpResponse> Trace(string url, string body)
        {
            CheckIfEntityExists();
            GCHandle handle;
            HttpResponse? data = null;
            var semaphore = new SemaphoreSlim(0, 1);

            unsafe
            {
                void ResolveTask(int statusCode, string body, IntPtr headerKeys, IntPtr headerValues, int headerSize)
                {
                    data = GetHttpResponse(statusCode, body, headerKeys, headerValues, headerSize);
                    semaphore.Release();
                }

                HttpResponseModuleDelegate resolveTask = ResolveTask;
                handle = GCHandle.Alloc(resolveTask);
                var urlPtr = MemoryUtils.StringToHGlobalUtf8(url);
                var bodyPtr = MemoryUtils.StringToHGlobalUtf8(body);
                Core.Library.Client.HttpClient_Trace(NativePointer, urlPtr, bodyPtr, resolveTask);
                Marshal.FreeHGlobal(urlPtr);
                Marshal.FreeHGlobal(bodyPtr);
            }

            await semaphore.WaitAsync();
            handle.Free();
            semaphore.Dispose();

            return data.Value;
        }
    }
}